### PR TITLE
DOM components as refs

### LINKED ISSUE #3223: Expose refs on DOM Components as Direct Handles to the actual DOM Node
My mental model is that we have a polyfill layer (`ReactDOMInput`) to support what the browser would support, and will support in the future once these ideas propagate through the standards. The fact that it is currently implemented as a wrapper should be an unobservable implementation detail, and perhaps it shouldn't be implemented as a wrapper.

As an upgrade path we would need to monkey patch existing legacy methods (`props`, `setProps`, `getDOMNode`) onto it as an upgrade path. These would work but immediately warn when used.

`setProps` will be replaced by some other imperative layer, or simply multiple calls to `React.render`.

`component.getDOMNode` is being replaced by `React.findDOMNode(component)` which can take either an arbitrary component or a DOM node directly. So you can use it on something that you're unsure about. In most cases, you can simply get rid of the call all together and use the DOM ref directly.

### PR DESCRIPTION
**Pull request by spicyj**
Fixes #3223.

### CODE REVIEW

File: src/isomorphic/classic/element/ReactElementValidator.js [hunk 1]
```diff
@@ -377,18 +377,10 @@ function checkAndWarnForMutatedProps(element) {
...
-    // This has already warned. Don't throw.
+  var componentClass = element.type;
+  if (typeof componentClass !== 'function') {
     return;
   }
-  // Extract the component class from the element. Converts string types
-  // to a composite class which may have propTypes.
-  // TODO: Validating a string's propTypes is not decoupled from the
-  // rendering target which is problematic.
-  var componentClass = ReactNativeComponent.getComponentClassForElement(
```
Comment (sebmarkbage) [inline src/isomorphic/classic/element/ReactElementValidator.js]: This makes ReactNativeComponent unused. rm the require.

File: src/renderers/dom/client/ReactMount.js [hunk 2]
```diff
@@ -330,6 +334,17 @@ function unmountComponentFromNode(instance, container) {
 }
 
 /**
+ * Temporary (?) hack so that we can store all top-level pending updates on
+ * composites instead of having to worry about different types of components
+ * here.
+ */
```
Comment (sebmarkbage) [inline src/renderers/dom/client/ReactMount.js]: Note: We'll have to remember to deal with this in devtools etc.

**Comment by sebmarkbage** [comment]
Looks like you have some test failures. Probably need to fix that. Other than that it looks good. Hopefully we figure out something cleaner for the top level node.

**Comment by spicyj** [comment]
Test failures are just due to missing warnings/getDOMNode. Been working on today, almost done.

**Comment by sebmarkbage** [comment]
I think that `getDOMNode` and `props` only needs to be patched in if there is a ref so `getPublicInstance` is the place to do it, right? We don't need it if `findDOMNode` is used to get the instance for example.

**Comment by spicyj** [comment]
@sebmarkbage Yeah, just pushed the last commit. Want to look? Maybe there's a better way to do setProps (cf. `_topLevelWrapper`) but this works.

(One failing test, investigating now.)

File: src/renderers/dom/shared/ReactDOMComponent.js [hunk 3]
```diff
@@ -817,6 +940,41 @@ ReactDOMComponent.Mixin = {
...
+
+  getPublicInstance: function() {
+    if (!this._nodeWithLegacyProperties) {
+      var node = ReactMount.getNode(this._rootNodeID);
+
+      node._reactInternalComponent = this;
+      node.getDOMNode = legacyGetDOMNode;
+      node.isMounted = legacyIsMounted;
+      node.setState = legacySetStateEtc;
+      node.replaceState = legacySetStateEtc;
```
Comment (sebmarkbage) [inline src/renderers/dom/shared/ReactDOMComponent.js]: Do we need to patch in all of these? Only setProps/replaceProps was available before 0.13 and setState/replaceState is a complete noop so nobody is likely to use them.

**Comment by sebmarkbage** [comment]
nbd if the setProps stuff gets messy. We'll make a quick follow up release that kills setProps/getDOMNode.

**Comment by spicyj** [comment]
yolo

**Comment by fabiomcosta** [comment]
Can you help me understand what does `TopLevelWrapper` represents for ReactPerf?
It shows on `ReactDefaultPerf.printInclusive()` as the most expensive render.

**Comment by sebmarkbage** [comment]
It is a wrapper around the root so, inclusively, it include everything else. So it should be the most expensive (inclusive).

**Pull request merged**